### PR TITLE
Ablations

### PR DESCRIPTION
**Pull request by gabrieltseng**
| | Name | Description | Priority | Evals Running |
| --- | --- | --- | --- | --- |
| 1 | phase2.0_base_no_contrastive | Contrastive weight set to 0 | High | yes |
| 2 | phase2.0_base_random_masking | random masking | High |  |
| 3 | phase2.0_base_mae | MAE instead of contrastive loss | Normal | yes |
| 4 | phase2.0_base_random_target | Randomly reinitialize the target encoder | Normal | |
| 5 | phase2.0_base_patchdisc_loss | Patch disc (not modality patch disc) loss | Normal | |
| 6 | phase2.0_base_no_ag_maps | No ag maps (remove CDL, WorldCereal) | High | yes |
| 7 | phase2.0_base_no_maps | No maps (also remove WorldCover, CHM, OSM) | High | yes |
| 8 | phase2.0_base_no_maps_srtm | No maps or SRTM (also remove SRTM) | Normal | |
| 9 | phase2.0_base_no_maps_srtm_landsat | No maps or SRTM  or Landsat(also remove Landsat) | Normal | |
| 10 | phase2.0_base_no_maps_srtm_landsat_s1 | No maps or SRTM  or Landsat or S1 (also remove S1) | Normal | |
| 11 | phase2.0_base_ema | EMA actually running | Low | yes |

### CODE REVIEW

File: helios/nn/flexihelios.py [hunk 1]
```diff
@@ -246,8 +246,9 @@ def pool_instance_wise(self, pooling_type: PoolingType) -> Tensor:
             num_encoded_tokens = torch.sum(mask, -1, keepdim=True)
             logger.debug(f"num_encoded_tokens: {num_encoded_tokens}")
             if (num_encoded_tokens == 0).any():
+                zero_encoded = torch.argwhere(num_encoded_tokens == 0)
                 raise ValueError(
-                    f"num_encoded_tokens is 0 for some samples {num_encoded_tokens}"
+                    f"num_encoded_tokens is 0 for some samples {num_encoded_tokens}, input_shape: {self.sentinel2_l2a.shape}, mask at that index is {self.sentinel2_l2a_mask[zero_encoded]}"  # type: ignore
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply helios/nn/flexihelios.py]: todo: this error message was only updated for debugging and should be reverted before merging

File: olmoearth_pretrain/train/train_module/contrastive_latentmim.py [hunk 2]
```diff
@@ -161,6 +164,12 @@ def __init__(
         self.mae_loss = mae_loss_config.build() if mae_loss_config is not None else None
         if self.mae_loss is not None:
             self.total_loss_name = f"{self.total_loss_name}+{self.mae_loss.name}"
+        if reinit_targets:
+            if ema_decay != (0.0, 0.0):
+                logger.warning(
+                    "Applying EMA updates to a randomly initialized target encoder."
+                )
+            self.model.target_encoder.apply(self.model.target_encoder._init_weights)
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply olmoearth_pretrain/train/train_module/contrastive_latentmim.py:172]: this doesn't fail - currently running

File: helios/nn/latent_mim.py [hunk 3]
```diff
@@ -126,6 +126,7 @@ class LatentMIMConfig(Config):
     encoder_config: Config
     decoder_config: Config
     reconstructor_config: Config | None = None
+    reinitialize_targets: bool = False
```
Comment (yawenzzzz) [inline helios/nn/latent_mim.py]: `reinitialize_targets` seems not been used, what is it for?

File: scripts/2025_10_02_phase2/script.py
Comment (yawenzzzz) [inline scripts/2025_10_02_phase2/script.py:1]: Suggestion: better to not changing `script.py` in phase 2, because our large/base/tiny/nano runs imports things from `script.py`. It may break the eval runs later. You can have another script just for the ablations.

File: scripts/2025_10_02_phase2/ablations/base_launch_ablations.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# no landsat
+python scripts/2025_10_02_phase2/base.py launch phase2.0_base_no_maps_srtm_landsat ai2/ceres-cirrascale  --launch.clusters='[ai2/jupiter-cirrascale-2,ai2/ceres-cirrascale]' --launch.priority=normal --common.training_modalities='[sentinel2_l2a,sentinel1]' --train_module.masking_config.strategy_config.only_decode_modalities='[]' --trainer.callbacks.wandb.project=2025_10_08_phase2_ablations
+# s2 only (no s1)
+python scripts/2025_10_02_phase2/base.py launch phase2.0_base_no_maps_srtm_landsat_s1_v3 ai2/ceres-cirrascale  --launch.clusters='[ai2/jupiter-cirrascale-2,ai2/ceres-cirrascale]' --launch.priority=urgent --common.training_modalities='[sentinel2_l2a]' --train_module.masking_config.strategy_config.only_decode_modalities='[]' --trainer.callbacks.wandb.project=2025_10_08_phase2_ablations
```
Comment (yawenzzzz) [inline scripts/2025_10_02_phase2/ablations/base_launch_ablations.sh:23]: Do we want to have a S2 + all maps? Wondering how much the maps can help in filling the gap in SSL

File: olmoearth_pretrain/train/masking.py [hunk 5]
```diff
@@ -305,6 +305,71 @@ def _create_random_mask(
...
+        num_not_missing_tokens = sum(not_missing_tokens)
+
+        if encode_ratio is None:
+            encode_ratio = self.encode_ratio
+        if decode_ratio is None:
+            decode_ratio = self.decode_ratio
+
+        encode_tokens = int(num_not_missing_tokens * encode_ratio)
+        decode_tokens = int(num_not_missing_tokens * decode_ratio)
+
```
Comment (yawenzzzz) [inline olmoearth_pretrain/train/masking.py:344]: Suggestion: adding check on the `num_not_missing_tokens`, if equals to 0, raise ValueError, if > 0, make sure encoder gets at least 1 tokens, we can do `max(1, int(num_not_missing_tokens * encode_ratio))` which guarantees that if only one token not missing, we always encode it. The decoder should be fine, in the patch discrimination loss, if one sample no tokens to decode, it will be skipped.

File: helios/train/masking.py [hunk 6]
```diff
@@ -305,6 +305,71 @@ def _create_random_mask(
...
+        encode_tokens = int(num_not_missing_tokens * encode_ratio)
+        decode_tokens = int(num_not_missing_tokens * decode_ratio)
+
+        # I assume if you are using this function its to make sure
+        # we have at least one encode and one decode token.
+        if encode_tokens == 0:
+            raise ValueError(f"No encode tokens for input shape {mask.shape}")
+        if decode_tokens == 0:
+            raise ValueError(f"No encode tokens for input shape {mask.shape}")
+
```
Comment (yawenzzzz) [inline helios/train/masking.py]: The `decode_tokens` can be skipped I think? In rare cases we will only see 1 token not missing, which will be given to the encoder, so `decode_tokens` will be 0 but it's fine.

File: scripts/2025_10_02_phase2/script.py
Comment (gabrieltseng) [reply scripts/2025_10_02_phase2/script.py:1]: yes - this was just temporary to make my life a bit easier. I'll change back before merging

Comment (gabrieltseng) [reply scripts/2025_10_02_phase2/script.py:1]: updated

File: helios/nn/latent_mim.py (hunk 3 above)
Comment (gabrieltseng) [reply helios/nn/latent_mim.py]: oops. removed

File: scripts/2025_10_02_phase2/ablations/base_launch_ablations.sh (hunk 4 above)
Comment (gabrieltseng) [reply scripts/2025_10_02_phase2/ablations/base_launch_ablations.sh:23]: IMO we already have the no-maps ablation which covers this.

File: helios/train/masking.py (hunk 6 above)
Comment (gabrieltseng) [reply helios/train/masking.py]: Removed these value errors - they were just to confirm a debugging hypothesis.

File: olmoearth_pretrain/train/masking.py [hunk 7]
```diff
@@ -1102,9 +1164,46 @@ def apply_bandset_mask_rules(
...
+                    )
+        for i in no_decoded_indices:
+            for key, val in masked_batch_dict.items():
+                if key.endswith("mask"):
+                    modality_mask = val[i]
+                    modality_name = "_".join(key.split("_")[:-1])
+                    modality_spec = Modality.get(modality_name)
+                    masked_batch_dict[key][i] = self._random_fill_unmasked(
+                        modality_mask, modality_spec, patch_size
+                    )
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply olmoearth_pretrain/train/masking.py:1218]: > The _random_fill_unmasked method is designed to operate on a mask for a single bandset.

no its not

File: olmoearth_pretrain/train/masking.py (hunk 5 above)
Comment (gabrieltseng) [reply olmoearth_pretrain/train/masking.py:344]: I won't raise a value error since we call this on individual band sets where its plausible not_missing_tokens == 0 but things are still okay.

I will make the second change where we make sure if there's only one not-missing, it gets assigned to encode (otherwise both encode and decode get 0 because of the `int()` call).

File: olmoearth_pretrain/train/masking.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply olmoearth_pretrain/train/masking.py:1218]: we know

File: olmoearth_pretrain/train/callbacks/evaluator_callback.py [hunk 8]
```diff
@@ -283,8 +283,24 @@ def _check_supported_modalities(self, evaluator: DownstreamEvaluator) -> bool:
...
+                f"model has supported_modalities: {self.trainer.train_module.model.supported_modalities}"
+            )
             return self.trainer.train_module.model.supported_modalities
+        elif hasattr(self.trainer.train_module.model, "encoder"):
+            logger.info("model has encoder")
+            if hasattr(self.trainer.train_module.model.encoder, "supported_modalities"):
+                logger.info(
+                    f"encoder has supported_modalities: {self.trainer.train_module.model.encoder.supported_modality_names}"
+                )
+                return self.trainer.train_module.model.encoder.supported_modality_names
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply olmoearth_pretrain/train/callbacks/evaluator_callback.py:394]: fixed

File: scripts/2025_10_02_phase2/script.py [hunk 9]
```diff
@@ -283,7 +283,7 @@ def build_trainer_config(common: CommonComponents) -> TrainerConfig:
             norm_stats_from_pretrained=True,
             input_modalities=[Modality.LANDSAT.name],
             input_layers=["landsat"],
-            eval_interval=Duration.steps(20000),
+            eval_interval=Duration.steps(10),
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply scripts/2025_10_02_phase2/script.py]: ctrl+z

File: olmoearth_pretrain/train/callbacks/evaluator_callback.py [hunk 10]
```diff
@@ -356,6 +365,11 @@ def post_step(self) -> None:
             )
             if self.step <= 1 or self.step % eval_interval_steps != 0:
                 continue
+            if not self._check_supported_modalities(evaluator):
+                logger.info(
+                    f"Skipping {evaluator.evaluation_name} because it requires a modality that is not supported by the model"
+                )
```
Comment (gabrieltseng) [inline olmoearth_pretrain/train/callbacks/evaluator_callback.py:473]: We now check the modalities in the in-loop callback too, since some ablations only have a subset of modalities.

File: olmoearth_pretrain/train/masking.py [hunk 11]
```diff
@@ -1102,17 +1106,116 @@ def apply_bandset_mask_rules(
...
+        for i in no_decoded_indices:
+            for key, val in masked_batch_dict.items():
+                if key.endswith("mask"):
+                    modality_mask = val[i]
+                    modality_name = "_".join(key.split("_")[:-1])
+                    modality_spec = Modality.get(modality_name)
+                    masked_batch_dict[key][i] = self._random_fill_unmasked(
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply olmoearth_pretrain/train/masking.py:1216]: thanks cursor!
